README.md

# Ergogen    [![Badge License]][License]

***Keyboard Generator*** *build on* ***[Absolem Keyboard]***.

<br>

<img
    src = 'docs/Showcase.png'
    width = 300
    align = right
/>

The project aims to provide a common configuration format to <br>
describe ***ergonomic*** 2D layouts and generate automatic plates, <br>
cases, as well as un-routed PCBs for them.

<br>
<br>

<div align = center>

---

[![Button Demo]][Demo]   
[![Button Getting Started]][Getting Started]   
[![Button Documentation]][Documentation]   
[![Button Discord]][Discord]

---

</div>

<br>
<br>

## Contributions

Feature ideas, documentation improvements, <br>
examples, tests, or pull requests welcome!

*Get in touch on our **[Discord]**, and we can definitely* <br>
*find something you can help with, if you'd like to.*

<br>

<!----------------------------------------------------------------------------->

[Badge License]: https://img.shields.io/badge/License-MIT-yellow.svg?style=for-the-badge

[Absolem Keyboard]: https://zealot.hu/absolem
[Getting Started]: docs/Getting%20Started.md
[Documentation]: https://docs.ergogen.xyz
[Discord]: https://discord.gg/nbKcAZB
[Demo]: https://ergogen.xyz

[License]: LICENSE


<!--------------------------------{ Buttons }---------------------------------->

[Button Getting Started]: https://img.shields.io/badge/Getting_Started-F94877?style=for-the-badge&logoColor=white&logo=GitHub
[Button Documentation]: https://img.shields.io/badge/Documentation-1793D1?style=for-the-badge&logoColor=white&logo=GitBook
[Button Discord]: https://img.shields.io/badge/Discord-5865F2?style=for-the-badge&logoColor=white&logo=Discord
[Button Demo]: https://img.shields.io/badge/Demo-37a779?style=for-the-badge&logoColor=white&logo=AppleArcade


changelog.md
# 4.0.0 (2022-??-??)

## :boom: Breaking Changes
## :seedling: Enhancements
## :skull: Bugfixes
## :wrench: Other

* fdc2adf - Dependency and roadmap housecleaning (2022-02-27)
* e13927d - Make reference dumping easier during testing (2022-02-27)
* b928cbd - Implement a simple autobind (2022-02-27)
* d23bd71 - Support orient/rotate towards other points (2022-02-27)
* 6dc6b5d - Anchor recursivization (2022-02-27)
* b8c71be - Improve readme (2022-01-24)
* cff15dd - Add missing bbox test (2022-01-23)
* d512983 - More flexible semver handling (2022-01-23)
* e0f5c91 - Fix rollup build warning (2022-01-23)
* 06d2ae4 - Switch to handrolled semver implementation (2022-01-23)
* 2b98b50 - Support string templating for key-level attributes (2022-01-22)
* 6225013 - Indicate dev version in package.json (2022-01-20)
* 7f5e5e7 - Roadmap update (2022-01-18)
* 28d076e - Remove accidentally added folder (2022-01-16)
* d6f8323 - Outlines rewrite actually done (2022-01-16)
* 4844a04 - Outlines rewrite theoretically done (2022-01-11)
* df7b76c - Outlines rewrite, part 2 (2022-01-10)
* 6504b2b - Outlines rewrite in progress (2022-01-09)
* bd6b5a0 - Add filter tests (2022-01-08)
* a7f333c - Filter implementation progress (2022-01-03)
* 534ac4b - Filter implementation started (2021-12-26)
* e48631f - Add key-level orient (2021-12-25)
* f2bd0d2 - Roadmap update (2021-12-18)
* 2c18902 - Simplify the names in individual point "zones" (2021-12-18)
* 9ee099b - Minor test adjustments (2021-12-18)
* 0ed29e7 - Place rectangles by their centers (2021-12-18)
* 73045e4 - Move column-level attributes to key-level (2021-12-18)







# 3.1.2 (2022-01-20)

* 86c00eb - (tag: v3.1.2, origin/master, backup/master, master) Bump patch version (2022-01-20)
* 3570be2 - Delete old docs (2022-01-20)
* f326594 - De-dupe readme, point usage to docs (2022-01-20)



# 3.1.1 (2022-01-14)

* 47d15c0 - (tag: v3.1.1) Hotfix parameterization (2022-01-14)


# 3.1.0 (2021-12-15)

* 488fba0 - (tag: v3.1.0) 3.1.0 (2021-12-15)
* f16ccd4 - Refresh roadmap (2021-12-15)
* 21cfc1f - Allow glue opt-out, even when bound (2021-12-15)
* c46a0bc - Add U for 19.05 spacing, closes #48 (2021-12-15)
* 3745cf3 - Parameterize PCB component reference hiding (2021-12-15)
* 13052cd - PCB metadata propagation (2021-12-15)
* 90db02f - Roadmap update (2021-12-15)
* 3e0dc84 - Support semver checks (2021-12-15)


# 3.0.3 (2021-12-13)

* d5ebd70 - (tag: v3.0.3) 3.0.3 (2021-12-13)
* e45e563 - Further dependency restructure (2021-12-13)



# 3.0.2 (2021-12-13)

* 96c7caa - (tag: v3.0.2) 3.0.2 (2021-12-13)
* 16c9243 - Move custom dependencies to ergogen org + fix metadata (2021-12-13)
* 5046667 - Suppress npm warnings (2021-12-13)
* e2535da - Update dependencies (2021-12-13)
* d193af6 - Add fresh TODO items, change to gitflow (2021-12-12)

# 3.0.1 (2021-12-12)

* 9a1f9ec - (tag: v3.0.1) 3.0.1 (2021-12-12)
* 16f6b81 - Fillet all outline chains, closes #34 (2021-12-12)
* 8490f14 - Add const in scrollwheel footprint (2021-08-16) <Michael van Eerd>
* 21e173e - Make spread: 0 possible (2021-08-13) <Michael van Eerd>
* 38a8fc5 - Anchor affect bugfix, partially fixes #33 (2021-12-12)
* 483f214 - Array unnest bugfix, fixes #50 (2021-12-12)
* e58b80c - Redirect readme links to new domain (2021-11-27)
* e29613d - Dump accumulated TODOs (2021-11-27)
* 0ec16db - Minor fix for mirrored anchor rotations (2021-11-27)
* a593f61 - Fixed rotary encoder footprint (2021-11-07) <Albert Engelbrecht>
* 68102e8 - Fix ReferenceError: *_unit is not defined (2021-09-03) <Michael van Eerd>
* bf2dee6 - docs(readme): fix typo in web ui url link (2021-08-08) <Michael van Eerd>
* ad5ac9c - Visualization fix (2021-07-25)

# 3.0.0 (2021-07-21)

* 0658057 - (tag: v3.0.0) 3.0.0 (2021-07-21)
* fa850a6 - YAML-ification and minor output restructure (2021-07-21)
* 282e117 - Web release prep (2021-07-21)
* 67e01ed - Base key-level defaults around `u` (2021-07-18)
* 9bdcd36 - Some outline testing (2021-07-18)
* 26128f8 - Support full anchors in outline shapes (2021-07-18)
* 677fae0 - Add roadmap (2021-07-18)
* b159cfa - Further testing (2021-07-18)
* 4105718 - Complete CLI testing (2021-07-17)
* 4d88dac - End-to-end CLI tests (2021-07-17)
* 1f3ecb5 - Improved interface testing (2021-07-16)
* 7d841b2 - Dependency updates (2021-07-16)
* 98d7f66 - Make old jscad stuff audit friendly (2021-07-16)
* 58cadde - Minor case fix + test (2021-07-16)
* f955aac - Support KLE to PCB metadata and nets (2021-07-16)
* cd0ae6d - Smarter dump test switch (2021-07-16)
* b43b719 - Auto-debug when output would be empty (2021-07-16)
* bc75781 - Adjust tests for async interface (2021-07-15)
* 1cb9fdc - Basic KLE support (2021-07-15)
* d09b3fd - Smarter dump switch for the test runner, reference adjustments (2021-07-14)
* 42a3e2d - CLI and output restructure, SVG/DXF/STL integration (2021-07-14)
* c49881c - Remove leftover debug statements (2021-07-11)
* c12e8d2 - Update footprints to match new structure (2021-07-11)
* fe30b91 - PCB net and parameter overhaul (2021-07-11)
* 452d7c1 - Units separated to their own block, and properly tested (2021-07-11)
* 71efdbe - Test switch overhaul (2021-07-11)
* ea12df2 - Fix typo (2021-07-11)
* 2425b06 - Package audit (2021-07-06)
* 2f6b9ab - Added new footprints for OLEDs, jumpers, and Omron B3F-4055 switches (2021-07-06) <tapioki>
* a431bc4 - Added documentation for switch footprints (2021-07-06) <tapioki>
* ee43a93 - Added keycaps option to display mx or choc keycaps on Dwgs.User, and fixed mx footprint (2021-06-12) <tapioki>
* 1ff7e07 - Added new switch footprints, updated reversibility functionality, integrated hotswap footprints, added orientation to promicro, alphabetized index, and added net and paramter documentation (2021-05-23) <tapioki>
* 3e78e6e - Add files via upload (2021-04-26) <tapioki>
* 5a8a66b - Add files via upload (2021-04-26) <tapioki>
* 9baae15 - Add unit test for assertions (2021-05-22)
* c7b86c7 - Add unit test for anchors (2021-05-22)

src/pcbs.js
const m = require('makerjs')
const a = require('./assert')
const prep = require('./prepare')
const anchor_lib = require('./anchor')

const kicad_prefix = `
(kicad_pcb (version 20171130) (host pcbnew 5.1.6)

  (page A3)
  (title_block
    (title KEYBOARD_NAME_HERE)
    (rev VERSION_HERE)
    (company YOUR_NAME_HERE)
  )

  (general
    (thickness 1.6)
  )

  (layers
    (0 F.Cu signal)
    (31 B.Cu signal)
    (32 B.Adhes user)
    (33 F.Adhes user)
    (34 B.Paste user)
    (35 F.Paste user)
    (36 B.SilkS user)
    (37 F.SilkS user)
    (38 B.Mask user)
    (39 F.Mask user)
    (40 Dwgs.User user)
    (41 Cmts.User user)
    (42 Eco1.User user)
    (43 Eco2.User user)
    (44 Edge.Cuts user)
    (45 Margin user)
    (46 B.CrtYd user)
    (47 F.CrtYd user)
    (48 B.Fab user)
    (49 F.Fab user)
  )

  (setup
    (last_trace_width 0.25)
    (trace_clearance 0.2)
    (zone_clearance 0.508)
    (zone_45_only no)
    (trace_min 0.2)
    (via_size 0.8)
    (via_drill 0.4)
    (via_min_size 0.4)
    (via_min_drill 0.3)
    (uvia_size 0.3)
    (uvia_drill 0.1)
    (uvias_allowed no)
    (uvia_min_size 0.2)
    (uvia_min_drill 0.1)
    (edge_width 0.05)
    (segment_width 0.2)
    (pcb_text_width 0.3)
    (pcb_text_size 1.5 1.5)
    (mod_edge_width 0.12)
    (mod_text_size 1 1)
    (mod_text_width 0.15)
    (pad_size 1.524 1.524)
    (pad_drill 0.762)
    (pad_to_mask_clearance 0.05)
    (aux_axis_origin 0 0)
    (visible_elements FFFFFF7F)
    (pcbplotparams
      (layerselection 0x010fc_ffffffff)
      (usegerberextensions false)
      (usegerberattributes true)
      (usegerberadvancedattributes true)
      (creategerberjobfile true)
      (excludeedgelayer true)
      (linewidth 0.100000)
      (plotframeref false)
      (viasonmask false)
      (mode 1)
      (useauxorigin false)
      (hpglpennumber 1)
      (hpglpenspeed 20)
      (hpglpendiameter 15.000000)
      (psnegative false)
      (psa4output false)
      (plotreference true)
      (plotvalue true)
      (plotinvisibletext false)
      (padsonsilk false)
      (subtractmaskfromsilk false)
      (outputformat 1)
      (mirror false)
      (drillshape 1)
      (scaleselection 1)
      (outputdirectory ""))
  )
`

const kicad_suffix = `
)
`

const kicad_netclass = `
  (net_class Default "This is the default net class."
    (clearance 0.2)
    (trace_width 0.25)
    (via_dia 0.8)
    (via_drill 0.4)
    (uvia_dia 0.3)
    (uvia_drill 0.1)
    __ADD_NET
  )
`

const makerjs2kicad = exports._makerjs2kicad = (model, layer='Edge.Cuts') => {
    const grs = []
    const xy = val => `${val[0]} ${-val[1]}`
    m.model.walk(model, {
        onPath: wp => {
            const p = wp.pathContext
            switch (p.type) {
                case 'line':
                    grs.push(`(gr_line (start ${xy(p.origin)}) (end ${xy(p.end)}) (angle 90) (layer ${layer}) (width 0.15))`)
                    break
                case 'arc':
                    const arc_center = p.origin
                    const angle_start = p.startAngle > p.endAngle ? p.startAngle - 360 : p.startAngle
                    const angle_diff = Math.abs(p.endAngle - angle_start)
                    const arc_end = m.point.rotate(m.point.add(arc_center, [p.radius, 0]), angle_start, arc_center)
                    grs.push(`(gr_arc (start ${xy(arc_center)}) (end ${xy(arc_end)}) (angle ${-angle_diff}) (layer ${layer}) (width 0.15))`)
                    break
                case 'circle':
                    const circle_center = p.origin
                    const circle_end = m.point.add(circle_center, [p.radius, 0])
                    grs.push(`(gr_circle (center ${xy(circle_center)}) (end ${xy(circle_end)}) (layer ${layer}) (width 0.15))`)
                    break
                default:
                    throw new Error(`Can't convert path type "${p.type}" to kicad!`)
            }
        }
    })
    return grs.join('\n')
}

const footprint_types = require('./footprints')

exports.inject_footprint = (name, fp) => {
    footprint_types[name] = fp
}

const footprint = exports._footprint = (config, name, points, point, net_indexer, component_indexer, units, extra) => {

    if (config === false) return ''

    // config sanitization
    a.unexpected(config, name, ['type', 'anchor', 'nets', 'anchors', 'params'])
    const type = a.in(config.type, `${name}.type`, Object.keys(footprint_types))
    let anchor = anchor_lib.parse(config.anchor || {}, `${name}.anchor`, points, point)(units)
    const nets = a.sane(config.nets || {}, `${name}.nets`, 'object')()
    const anchors = a.sane(config.anchors || {}, `${name}.anchors`, 'object')()
    const params = a.sane(config.params || {}, `${name}.params`, 'object')()

    // basic setup
    const fp = footprint_types[type]
    const parsed_params = {}

    // connecting other, non-net, non-anchor parameters
    parsed_params.param = {}
    for (const [param_name, param_value] of Object.entries(prep.extend(fp.params || {}, params))) {
        let value = param_value
        if (a.type(value)() == 'string' && value.startsWith('=') && point) {
            const indirect = value.substring(1)
            value = point.meta[indirect]
            if (value === undefined) {
                throw new Error(`Indirection "${name}.params.${param}" --> "${point.meta.name}.${indirect}" to undefined value!`)
            }
        }
        parsed_params.param[param_name] = value
    }

    // reference
    const component_ref = parsed_params.ref = component_indexer(parsed_params.param.class || '_')
    parsed_params.ref_hide = extra.references ? '' : 'hide'

    // footprint positioning
    parsed_params.at = `(at ${anchor.x} ${-anchor.y} ${anchor.r})`
    parsed_params.rot = anchor.r
    parsed_params.xy = (x, y) => {
        const new_anchor = anchor_lib.parse({
            shift: [x, -y]
        }, '_internal_footprint_xy', points, anchor)(units)
        return `${new_anchor.x} ${-new_anchor.y}`
    }

    // connecting nets
    parsed_params.net = {}
    for (const [net_name, net_value] of Object.entries(prep.extend(fp.nets || {}, nets))) {
        let net = a.sane(net_value, `${name}.nets.${net_name}`, 'string')()
        if (net.startsWith('=') && point) {
            const indirect = net.substring(1)
            net = point.meta[indirect]
            net = a.sane(net, `${name}.nets.${net_name} --> ${point.meta.name}.${indirect}`, 'string')()
        }
        // Prevent duplicate net names between the main board and its mirror
        if (point.meta.name.startsWith("mirror_")) {
            net = `mirror_${net}`
        }
        const index = net_indexer(net)
        parsed_params.net[net_name] = {
            name: net,
            index: index,
            str: `(net ${index} "${net}")`
        }
    }

    // allowing footprints to add dynamic nets
    parsed_params.local_net = suffix => {
        const net = `${component_ref}_${suffix}`
        const index = net_indexer(net)
        return {
            name: net,
            index: index,
            str: `(net ${index} "${net}")`
        }
    }

    // parsing anchor-type parameters
    parsed_params.anchors = {}
    for (const [anchor_name, anchor_config] of Object.entries(prep.extend(fp.anchors || {}, anchors))) {
        let parsed_anchor = anchor_lib.parse(anchor_config || {}, `${name}.anchors.${anchor_name}`, points, anchor)(units)
        parsed_anchor.y = -parsed_anchor.y
        parsed_params.anchors[anchor_name] = parsed_anchor
    }

    return fp.body(parsed_params)
}

exports.parse = (config, points, outlines, units) => {

    const pcbs = a.sane(config.pcbs || {}, 'pcbs', 'object')()
    const results = {}

    for (const [pcb_name, pcb_config] of Object.entries(pcbs)) {

        // config sanitization
        a.unexpected(pcb_config, `pcbs.${pcb_name}`, ['outlines', 'footprints', 'references'])
        const references = a.sane(pcb_config.references || false, `pcbs.${pcb_name}.references`, 'boolean')()

        // outline conversion
        if (a.type(pcb_config.outlines)() == 'array') {
            pcb_config.outlines = {...pcb_config.outlines}
        }
        const config_outlines = a.sane(pcb_config.outlines || {}, `pcbs.${pcb_name}.outlines`, 'object')()
        const kicad_outlines = {}
        for (const [outline_name, outline] of Object.entries(config_outlines)) {
            const ref = a.in(outline.outline, `pcbs.${pcb_name}.outlines.${outline_name}.outline`, Object.keys(outlines))
            const layer = a.sane(outline.layer || 'Edge.Cuts', `pcbs.${pcb_name}.outlines.${outline_name}.outline`, 'string')()
            kicad_outlines[outline_name] = makerjs2kicad(outlines[ref], layer)
        }

        // making a global net index registry
        const nets = {"": 0}
        const net_indexer = net => {
            if (nets[net] !== undefined) return nets[net]
            const index = Object.keys(nets).length
            return nets[net] = index
        }
        // and a component indexer
        const component_registry = {}
        const component_indexer = _class => {
            if (!component_registry[_class]) {
                component_registry[_class] = 0
            }
            component_registry[_class]++
            return `${_class}${component_registry[_class]}`
        }

        const footprints = []

        // key-level footprints
        for (const [p_name, point] of Object.entries(points)) {
            for (const [f_name, f] of Object.entries(point.meta.footprints || {})) {
                footprints.push(footprint(f, `${p_name}.footprints.${f_name}`, points, point, net_indexer, component_indexer, units, {references}))
            }
        }

        // global one-off footprints
        if (a.type(pcb_config.footprints)() == 'array') {
            pcb_config.footprints = {...pcb_config.footprints}
        }
        const global_footprints = a.sane(pcb_config.footprints || {}, `pcbs.${pcb_name}.footprints`, 'object')()
        for (const [gf_name, gf] of Object.entries(global_footprints)) {
            footprints.push(footprint(gf, `pcbs.${pcb_name}.footprints.${gf_name}`, points, undefined, net_indexer, component_indexer, units, {references}))
        }

        // finalizing nets
        const nets_arr = []
        const add_nets_arr = []
        for (const [net, index] of Object.entries(nets)) {
            nets_arr.push(`(net ${index} "${net}")`)
            add_nets_arr.push(`(add_net "${net}")`)
        }

        const netclass = kicad_netclass.replace('__ADD_NET', add_nets_arr.join('\n'))
        const nets_text = nets_arr.join('\n')
        const footprint_text = footprints.join('\n')
        const outline_text = Object.values(kicad_outlines).join('\n')
        const personalized_prefix = kicad_prefix
            .replace('KEYBOARD_NAME_HERE', pcb_name)
            .replace('VERSION_HERE', config.meta && config.meta.version || 'v1.0.0')
            .replace('YOUR_NAME_HERE', config.meta && config.meta.author || 'Unknown')
        results[pcb_name] = `
            ${personalized_prefix}
            ${nets_text}
            ${netclass}
            ${footprint_text}
            ${outline_text}
            ${kicad_suffix}
        `
    }

    return results
}
